### PR TITLE
test: add engine tree test, FCU triggers reorg with all the blocks present

### PR DESCRIPTION
**Pull request by fgimenez**
Towards #9749 

The test raised the lack of some events not being emitted, they are added now to `EngineApiTreeHandlerImpl`:  `EngineApiEvent::BeaconConsensus(BeaconConsensusEngineEvent::ForkBlockAdded)`, `EngineApiEvent::BeaconConsensus(BeaconConsensusEngineEvent::CanonicalBlockAdded)` and `EngineApiEvent::BeaconConsensus(BeaconConsensusEngineEvent::CanonicalChainCommitted)`.

Besides the test and the changes above, the PR also includes:
* Changes in `TestBlockBuilder` to get the execution outcome of a given block
* Changes in `MockEthProvider` to include a state root store, and use it in the `StateRootProvider` impl methods.

### CODE REVIEW

File: crates/engine/tree/src/tree/mod.rs [hunk 1]
```diff
@@ -1132,6 +1136,8 @@ where
     /// This is invoked on a valid forkchoice update, or if we can make the target block canonical.
     fn on_canonical_chain_update(&mut self, chain_update: NewCanonicalChain) {
         trace!(target: "engine", new_blocks = %chain_update.new_block_count(), reorged_blocks =  %chain_update.reorged_block_count() ,"applying new chain update");
+        let start = Instant::now();
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs:1141]: this is great, we def also need this for upcoming metrics

File: crates/engine/tree/src/tree/mod.rs [hunk 2]
```diff
@@ -1303,7 +1317,16 @@ where
 
         self.state.tree_state.insert_executed(executed);
 
+        // emit insert event
+        let engine_event = if self.state.tree_state.is_fork(&block.block) {
+            BeaconConsensusEngineEvent::ForkBlockAdded(sealed_block)
+        } else {
+            BeaconConsensusEngineEvent::CanonicalBlockAdded(sealed_block, start.elapsed())
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs:1326]: yes exactly

**Review by mattsse — APPROVED** [review]
very clean!

**Comment by fgimenez** [comment]
I've added a new test that exercises extending the canonical chain and expecting `CanonicalBlockAdded` events and this showed some issues in `is_fork` method, should be fixed now

**Review by mattsse — APPROVED** [review]
nice

**Pull request merged**